src/imitation/util/networks.py

```diff
@@ -165,26 +165,67 @@ def update_stats(self, batch: th.Tensor) -> None:
             batch: A batch of data to use to update the running mean and variance.
         """
         b_size = batch.shape[0]
+        # Shuffle the batch since we don't don't want to bias the mean
+        # towards data that appears latter in the batch
+        perm = th.randperm(b_size)
+        batch = batch[perm]
+        alpha = 1 - self.decay
 
         if self.count == 0:
-            self.running_mean = th.mean(batch, dim=0)
+            # geometric progession of decay: decay^(N-1),...,decay,1
+            weights = th.vander(th.tensor([self.decay]), N=b_size).T
+            weights[1:] *= alpha
+            self.running_mean = th.sum(weights * batch, dim=0)
             if b_size > 1:
-                self.running_var = th.var(batch, dim=0)
+                self.running_var = th.sum(weights * (batch**2), dim=0)
+                self.running_var -= self.running_mean**2
         else:
-            # Shuffle the batch since we don't don't want to bias the mean
-            # towards data that appears latter in the batch
-            perm = th.randperm(b_size)
-
-            alpha = 1 - self.decay
-
             for i in range(b_size):
-                diff = batch[perm[i], ...] - self.running_mean
+                diff = batch[i, ...] - self.running_mean
                 incr = alpha * diff
                 self.running_mean += incr
                 self.running_var = self.decay * (self.running_var + diff * incr)
 
         self.count += b_size
 
+    def update_stats(self, batch: th.Tensor) -> None:
+        """Update `self.running_mean` and `self.running_var` in batch mode.
+
+        Args:
+            batch: A batch of data to use to update the running mean and variance.
+        """
+        b_size = batch.shape[0]
+        batch = batch.reshape(b_size, -1)
+        # Shuffle the batch since we don't don't want to bias the mean
+        # towards data that appears latter in the batch
+        perm = th.randperm(b_size)
+        batch = batch[perm]
+        # geometric progession of decay: decay^(N-1),...,decay,1
+        weights = th.vander(th.tensor([self.decay]), N=b_size).T
+        alpha = 1 - self.decay
+        if self.count == 0:
```

Comment:
FWIW using a simple mean/variance on the first batch seemed OK to me. All data in the batch we can treat as having arrived at the same time, so there's no reason to weight some data in the batch more than others. This is effectively what the randomization is trying to prevent (by not *consistently* weighting some indices more than others), but for the first batch a simple average also works for this.

However I don't feel particularly strongly about this either way, it does seem nice to be consistent. Any thoughts @levmckinney?

Reply:
I updated the code for the first batch because of consistency reasons. 

Randomizing the batch before EMA updates did feel weird to me since randomizing counters with what we want to do with EMA (giving more weight to recently seen examples). A better approach could be to assign weights batch-wise. That is, give more weight to recently seen batches (by using the exponential decay) while keeping the same weight for all examples within a batch. For example, if the batch size is N and we have seen 4 batches till now, the weights for all the examples would be (recent examples on right):
(d^3)/N, ..., (d^3)/N, a(d^2)/N, ..., a(d^2)/N, ad/N, ..., ad/N, a/N, ..., a/N (weights sum to 1)
where I'm using ... to represent iteration over a batch of N examples (having equal weight).

This algorithm would be equivalent to doing EMA updates over the batch mean and EMA updates with the batch variance.

If this sounds better to you too, then I can implement this algorithm which wouldn't require randomization within a batch. The code would also be considerably easier to understand.

Reply:
> This algorithm would be equivalent to doing EMA updates over the batch mean and EMA updates with the batch variance.

I'm fairly sure this is false or at least not true in all cases. 

First this algorithm is Ill defined when we have a batch size of 1. This is actually quite plausible when we are doing RL training in an expensive environment. Second, I tried to do something along these lines before. If I'm understanding your notation correctly, you're proposing something like the following,

Data $\{x_{k,j}\}$ first index is for batch second index is for element in batch.
Batch size: $b$
Number of batches:  $m$
Decay rate $\gamma$ and learning rate $\alpha = 1 - \gamma$
Let $w_{m, k} = \begin{cases} \frac{\gamma^{m - k}\alpha}{b} , 1 \leq k \leq m \\ \frac{\gamma^m}{b}, k = 0\end{cases}$  These two indices represent the largest batch index and the batch index.



$\mu_{m} = \mathbb E_{m}\left[x\right]$
$\sigma^2_{m} = \mathbb E_{m} [(x - \mu_{bm})^2]$

Now we can show that $\sigma^2_{m} = \mathbb E_{m} [(x - \mu_{bm})^2]$ is not the same as the EMA norm of the sample variances.

Consider the following counter example we have two batches {(0, 0), (1,1)}, $\gamma = \alpha = 1/2$.
Then computing $\sigma^2$ by doing EMA on the _batch sample variances_  gives $\sigma^2_1 = 0$. On the other hand, using the above approach gives $\sigma_{1,1} = \mathbb E_{1,1} [(x - \mu_{11})^2]$. First, we can compute $\mu_{1, 1} = w_{1,0}*0 + w_{1, 0}*0 + w_{1, 1}*1 + w_{1,1}*1 = \gamma\alpha = 1/4$. Thus finally, $\sigma_{11} = \mathbb E_{11} [(x - \mu_{11})^2] = \mathbb E_{11} [(x - 1/4)^2]  = w_{02}*\frac{1}{16} + w_{02}*\frac{1}{16} + w_{12}\frac{9}{16} + w_{12}*\frac{9}{16} = \frac{\gamma}{16} + \frac{\alpha\gamma 9}{16} =  \frac{11}{64}$

Reply:
Coming back to this again I think I've finally figured out the correct incremental algorithm. I've attached in bellow with a proof of correctness,

[Incremental EMA with batches.pdf](https://github.com/HumanCompatibleAI/imitation/files/9278587/Incremental.EMA.with.batches.pdf)

Reply:
Actually, I incorrectly wrote EMA over variance in the last comment. I meant to write that EMA would be equivalent as if we are doing updates on batch mean, and the EMV would be similar to what I have currently implemented in the code (the batch version of original incremental EMV), just that it would become a lot simpler if we take equal weights in a batch.

I looked at the pdf, and that is exactly what I was proposing. Just to confirm, I think in the pseudo-code, the subscript should be (k-1) instead of (m-1) everywhere.

Reply:
Yes it should be k - 1.

Reply:
This won't be numerically stable https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance.

Reply:
If we first multiply the examples by their weights and then compute the sum, that should be numerically stable, right? For simple mean and variance, the weights would be 1/N, where N is the total number of examples.

If the above is true, then the current code should be numerically stable.

Reply:
I believe this will be better

![image](https://user-images.githubusercontent.com/12176390/183529948-93df2557-64c7-4acd-92a9-0c73427380dc.png)

I'll get you the proof for this one soon

Reply:
Last one because I think it's nice!
![image](https://user-images.githubusercontent.com/12176390/183531003-ec2432ad-9385-4fe2-93ce-b2f1daaeba3f.png)
I'm a bit less confident this one is correct.

Reply:
Created an overleaf document https://www.overleaf.com/project/62f292899a3ec6a55359da7e